CanvasColorSpaceProposal.md

```diff
@@ -72,6 +70,7 @@ partial dictionary CanvasRenderingContext2DSettings {
 partial dictionary WebGLContextAttributes {
   CanvasColorSpaceEnum colorSpace = "srgb";
   CanvasStorageFormatEnum storageFormat = "unorm8";
+  boolean srgbColorEncoding = false;
```

Comment:
Probably unnecessary idea, but (if this is accepted) perhaps this could be expressed more explicitly with an enum involving "perceptually-linear" and "physically-linear"? I'm having trouble figuring out if this makes any sense though so feel free to ignore.

Reply:
This feature does not belong in this spec, and is only a source of confusion. It affects only WebGL, and should be exposed in a way that centers WebGL. It is meaningless for 2D canvas, and is not needed for WebGPU.

I have updated the patch to remove it.

This feature is useful, but should be added in a separate, WebGL-centric specification.

Reply:
Rather than split off this functionality into the WebGL spec, could we allow `unorm8` and `srgb-linear` to work with WebGL, and add a couple of caveats that readbacks via `readPixels` will yield slightly different results than expected?

The rendering and display portions should work the same as with the other rendering APIs. Most users won't encounter the behavioral difference with read-back pixel values, so it would be nice to request the canvas rendering context in the same way for WebGL as with e.g. 2D canvas.

Reply:
I think I see the source of everyone's disagreement here. EGL's definition of "color space" was a complete screw-up, and confuses everyone who reads it. Fortunately, EGL was the only thing that screwed this up, and everything else (Vulkan, Metal, D3D) didn't screw it up. So we can just throw EGL (well, this particular part of it) under the bus and all agree on everything!!!

Let's look at [Vulkan](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkColorSpaceKHR.html).

- The VkSurfaceFormatKHR specifies a VkFormat and a VkColorSpaceKHR independently.
- The color space can be VK_COLOR_SPACE_SRGB_NONLINEAR_KHR or VK_COLOR_SPACE_EXTENDED_SRGB_LINEAR_EXT. It can also be things like DISPLAY_P3_NONLINEAR, BT2020_LINEAR, HDR10_HLG, or a [bunch of other stuff](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkColorSpaceKHR.html). 
- The pixel format can be VK_FORMAT_R8G8B8A8_UNORM or VK_FORMAT_R8G8B8A8_SRGB (the later specifying that sRGB color encoding is desired). It can also be many other things like A2B10G10R10, R16G16B16A16_SFLOAT, and [many others](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkFormat.html).
- The "I want 8-bit physically based rendering" configuration is VK_COLOR_SPACE_SRGB_NONLINEAR_KHR with VK_FORMAT_R8G8B8A8_SRGB.
- Critically, the color space and the pixel format can be mixed-and-matched.

Let's look at [Metal](https://developer.apple.com/documentation/quartzcore/cametallayer?language=objc)

- The CAMetalLayer specifies a colorspace and pixelFormat separately
- Color space can be kCGColorSpaceSRGB or kCGColorSpaceLinearSRGB. It can also be AdobeRGB, DisplayP3, ITU_R2020, or another [whole bunch of stuff](https://developer.apple.com/documentation/coregraphics/cgcolorspace/color_space_names?language=objc).
- Pixel format can be MTLPixelFormatBGRA8Unorm or MTLPixelFormatBGRA8Unorm_sRGB (the latter having sRGB color encoding). It can also be the same bazillion [other formats](https://developer.apple.com/documentation/metal/mtlpixelformat?language=objc).
- The "I want 8-bit physically based rendering" configuration is kCGColorSpaceSRGB with MTLPixelFormatBGRA8Unorm_sRGB
- Critically, the color space and the pixel format can be mixed-and-matched.

Let's look at D3D.

- An IDXGISwapChain has a [DXGI_FORMAT](https://docs.microsoft.com/en-us/windows/win32/api/dxgiformat/ne-dxgiformat-dxgi_format) and can have a [DXGI_COLOR_SPACE_TYPE](https://docs.microsoft.com/en-us/windows/win32/api/dxgicommon/ne-dxgicommon-dxgi_color_space_type#see-also) set on it.
- The format can be DXGI_FORMAT_R8G8B8A8_UNORM or DXGI_FORMAT_R8G8B8A8_UNORM_SRGB, or again, many other things.
- These formats can be used with the color space DXGI_COLOR_SPACE_RGB_FULL_G22_NONE_P709 (which is sRGB). There are some other options, but it's pretty restrictive.
- The "I want 8-bit physically based rendering" configuration is DXGI_FORMAT_R8G8B8A8_UNORM_SRGB with DXGI_COLOR_SPACE_RGB_FULL_G22_NONE_P709
- This has the same structure as the other examples, but it doesn't allow as much mixing-and-matching. Most of the supported color spaces are for video.

Let's look at GLX and WGL.
- In 2008, [ARB_framebuffer_sRGB](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_framebuffer_sRGB.txt) was ratified.
 - It added the ability to enable/disable sRGB color encoding on the backbuffer using FRAMEBUFFER_SRGB
 - To be able to use that, one needs to specify WGL[GLX]_FRAMEBUFFER_SRGB_CAPABLE_ARB
 - This enables what people had been able to do with framebuffer objects bound to GL_SRGB8_ALPHA8 textures. The spec indicates the interactions with GL_FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING
- GLX never added color spaces (as far as I can see).

Then EGL came along and completely screwed up.
- The original sin is in 2014 with [KHR_gl_colorspace](https://www.khronos.org/registry/EGL/extensions/KHR/EGL_KHR_gl_colorspace.txt).
 - This spec wanted to do what was in ARB_framebuffer_sRGB. It has the same language about GL_FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING.
 - But it uses the word "color space" incorrectly. It allows specifying EGL_GL_COLORSPACE_KHR to be either SRGB or LINEAR. This has nothing to do with the color space of the framebuffer. It's a total misnomer.
 - Things get even worse. Consider the quote "[if set to] EGL_GL_COLORSPACE_SRGB_KHR, then [the color space is] a non-linear, perceptually uniform color space is assumed, with a corresponding GL_FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING value of GL_SRGB". This is also up a creek because sRGB is a perceptually uniform by a long shot. Something like [CIELUV](https://en.wikipedia.org/wiki/CIELUV) is. sRGB happens to be close-ish-maybe to perceptually uniform for grayscale values.
- Then EGL tried to add actual color spaces in 2017 with [EGL_EXT_gl_colorspace_scrgb](https://www.khronos.org/registry/EGL/extensions/EXT/EGL_EXT_gl_colorspace_scrgb.txt), [EXT_gl_colorspace_scrgb_linear](https://www.khronos.org/registry/EGL/extensions/EXT/EGL_EXT_gl_colorspace_scrgb_linear.txt), and [EXT_gl_colorspace_display_p3](https://www.khronos.org/registry/EGL/extensions/EXT/EGL_EXT_gl_colorspace_display_p3.txt), among [others](https://www.khronos.org/registry/EGL/extensions/EXT/).
 - These try to keep using the EGL_GL_COLORSPACE_KHR enum because it's the name for the thing that they want.
 - EGL_GL_COLORSPACE_SCRGB_EXT and EGL_GL_COLORSPACE_SCRGB_LINEAR_EXT are then intended only to be used with floating-point pixel formats.
 - I haven't managed to wrap my head around when to use P3, P3_LINEAR, and P3_PASSTHROUGH

The whole "srgb" versus "linear" thing is horribly confusing because EGL made it incredibly confusing. I honestly can't keep EGL straight in my head.

But, Vulkan, Metal, D3D, and even GLX, are very clear. So I think the goal should be to mirror the sane APIs, and their language. WebGPU fits this mix very well.

WebGL is a bit odd because we don't specify a pixel format (there's no "use GL_RGBA", there's "use alpha:true"). I think that the most consistent thing would be to have a srgbColorEncoding parameter (like the first version of this patch). It also mirrors the alpha behavior in that it can only be used with 8-bit textures (there is no RGB texture that is half-float -- you can't get rid of the alpha channel). I also think it should be in a separate specification (and I think it would be ratified faster than this spec, because it'd be so small).

(Apologies for any typos -- I have to run now, but I wanted to get this out earlier. I'll try wrapping my head around the various EGL options, so that at least I can document it here).

Reply:
Thanks for the in-depth background Chris.

What do you propose? In https://github.com/WICG/canvas-color-space/pull/42 - now reverted, because I merged it before Jeff had a chance to review it - "linear" versions of the various color spaces were added.

Do you suggest to factor out the transfer function somehow? Change the definitions of the color spaces and pixel formats? Something else?

Basically - it would be good to have some way to say at the JavaScript level that the user would like a GL_SRGB8_ALPHA8 texture attached as WebGL's default framebuffer. Ideally, reinterpreting whatever enums are defined in this spec, e.g. for WebGPU, and not bolted on separately to the WebGL spec. A little fudge factor can be added either to this spec or WebGL's to indicate that operations like `readPixels` won't return correct values like the WebGPU API would.

Reply:
> Do you suggest to factor out the transfer function somehow? Change the definitions of the color spaces and pixel formats? Something else?

We should factor out not not the transfer function, but the backbuffer's color encoding (aka FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING of framebuffer zero).
- We have to keep very clear that "color space" is completely orthogonal to FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING
- That's the mistake that EGL made that is causing so much confusion.
- FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING being set to SRGB causes a specific mathematical function to be applied to the gl_FragColor values after the blending stage of the pipeline.
  - That mathematical function happens to be the transformation from linear-sRGB space to sRGB space. This means that gl_FragColor is, effectively, in linear-sRGB space in that state.

I advocate that we be as explicit about it as possible, by adding a parameter WebGLContextAttributes.
- It could be "framebufferColorEncoding" and be set to "srgb" or "linear" (default "linear"), if we want to be consistent with OpenGL's terminology.
  - Going with "framebufferColorEncodingFunction" as "linear-to-srgb" or "identity" would be more precise, but breaking with tradition
- Alternatively, a boolean enableSrgbFramebufferColorEncoding and set to "true" or "false" (default "false"), if we want to try to avoid the bad "linear" terminology.
  - This is the best way to go.
  - This aligns with the way WebGL uses alpha:"false".
    - That is, in effect, specifying GL_RGB
    - It also only applies to 8 bit backbuffers
  - This mirrors GLX_FRAMEBUFFER_SRGB_CAPABLE_ARB from [ARB_framebuffer_sRGB](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_framebuffer_sRGB.txt)

When "enabled", by whatever, means, the transformation described in Section 17.3.9 of the [OpenGL 4.3 spec](https://www.khronos.org/registry/OpenGL/specs/gl/glspec43.core.pdf) would be applied.

> Ideally, reinterpreting whatever enums are defined in this spec, e.g. for WebGPU, and not bolted on separately to the WebGL spec.

Having a storage format of "unorm8-srgb" can be made to work. But a separate boolean is strictly superior. Four reasons
- It matches how alpha impacts the backbuffer's GL format
- It doesn't pollute the general-purpose CanvasStorageFormat enum (which is used by 2D canvas)
- It matches ARB_framebuffer_sRGB, which got this right
- It is orthogonal to the rest of the spec, and could be fast-tracked in its own spec.

> A little fudge factor can be added either to this spec or WebGL's to indicate that operations like readPixels won't return correct values like the WebGPU API would.

Having gone through the details of how EGL screwed this up, I come down on the side of a hard no to this. There is an integrity to separating FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING from the color space. EGL tried to munge them together and ended up with a mess. Vulkan/Metal/D3D did it right.

Reply:
I've edited my long-explanation-of-EGL comment to be more comprehensive and hopefully easier to follow.

I now very much understand the revulsion to the term "linear" and "srgb" -- the original EGL specs abused them horribly. But it's just the EGL specs (and one WGL spec). Everything else, and all of the newer APIs (Vulkan/MetalD3D) use these terms in a clear and coherent way. I don't think we should let KHR_gl_colorspace poison the standard terminology for us. (Though we should probably give ample pointers to ensure the misunderstandings that resulted in that spec are not repeated).

I've updated the patch to expose what I feel is the best API, for the reasons I described. As I mentioned before, I think that this should be in a separate spec, because it **is** orthogonal functionality and should be treated as such (and not doing so has dire consequences). But, I'm content to hitch it along for the ride here, in this form, if that makes everyone else happy.